pkg/operator/configobservation/etcd/observe_etcd.go

```diff
@@ -0,0 +1,89 @@
+package etcd
+
+import (
+	"fmt"
+	"net"
+	"reflect"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	"github.com/openshift/library-go/pkg/operator/configobserver"
+	"github.com/openshift/library-go/pkg/operator/events"
+
+	"github.com/openshift/cluster-openshift-apiserver-operator/pkg/operator/configobservation"
+	"github.com/openshift/cluster-openshift-apiserver-operator/pkg/operator/operatorclient"
+)
+
+const (
+	etcdEndpointNamespace = operatorclient.EtcdEndpointNamespace
+	etcdEndpointName      = operatorclient.EtcdEndpointName
+)
+
+// ObserveStorageURLs observes the storage config URLs. If there is a problem observing the current storage config URLs,
+// then the previously observed storage config URLs will be re-used.
+func ObserveStorageURLs(genericListers configobserver.Listers, recorder events.Recorder, currentConfig map[string]interface{}) (observedConfig map[string]interface{}, errs []error) {
+	listers := genericListers.(configobservation.Listers)
+	observedConfig = map[string]interface{}{}
+	storageConfigURLsPath := []string{"storageConfig", "urls"}
+
+	currentEtcdURLs, found, err := unstructured.NestedStringSlice(currentConfig, storageConfigURLsPath...)
+	if err != nil {
+		errs = append(errs, err)
+	}
+	if found {
+		if err := unstructured.SetNestedStringSlice(observedConfig, currentEtcdURLs, storageConfigURLsPath...); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	var etcdURLs []string
+	etcdEndpoints, err := listers.EndpointsLister.Endpoints(etcdEndpointNamespace).Get(etcdEndpointName)
+	if errors.IsNotFound(err) {
+		recorder.Warningf("ObserveStorageFailed", "Required %s/%s endpoint not found", etcdEndpointNamespace, etcdEndpointName)
+		errs = append(errs, fmt.Errorf("endpoints/host-etcd.kube-system: not found"))
+		return
+	}
+	if err != nil {
+		recorder.Warningf("ObserveStorageFailed", "Error getting %s/%s endpoint: %v", etcdEndpointNamespace, etcdEndpointName, err)
+		errs = append(errs, err)
+		return
+	}
+	for subsetIndex, subset := range etcdEndpoints.Subsets {
+		for addressIndex, address := range subset.Addresses {
+			ip := net.ParseIP(address.IP)
+			if ip == nil {
+				ipErr := fmt.Errorf("endpoints %s/%s: subsets[%v]addresses[%v].IP is not a valid IP address", etcdEndpointName, etcdEndpointNamespace, subsetIndex, addressIndex)
+				errs = append(errs, ipErr)
+				continue
+			}
+			if ip.To4() != nil {
+				etcdURLs = append(etcdURLs, fmt.Sprintf("https://%s:2379", ip))
+			} else {
+				etcdURLs = append(etcdURLs, fmt.Sprintf("https://[%s]:2379", ip))
+			}
+		}
+	}
+
+	if len(etcdURLs) == 0 {
+		emptyURLErr := fmt.Errorf("endpoints %s/%s: no etcd endpoint addresses found", etcdEndpointNamespace, etcdEndpointName)
```

Comment:
send event and cleanly set nothing to allow the default service hostname to work.

Reply:
> send event and cleanly set nothing to allow the default service hostname to work.

this will happen continuously until the etcd operator is updated

Reply:
Done